hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/RefreshMetaProcedure.java

```diff
@@ -0,0 +1,414 @@
+    LOG.debug("Found {} namespace directories in data dir", Arrays.stream(namespaceDirs).toList());
+
+    for (FileStatus nsDir : namespaceDirs) {
+      try {
+        List<RegionInfo> namespaceRegions = scanTablesInNamespace(fs, nsDir.getPath());
+        regions.addAll(namespaceRegions);
+        LOG.debug("Found {} regions in namespace {}", namespaceRegions.size(), nsDir.getPath().getName());
+      } catch (IOException e) {
+        LOG.error("Failed to scan namespace directory: {}", nsDir.getPath(), e);
+      }
+    }
+    LOG.info("Scanned backing storage and found {} regions", regions.size());
+    return regions;
+  }
+
+  private List<RegionInfo> scanTablesInNamespace(FileSystem fs, Path namespacePath) throws IOException {
+    LOG.debug("Scanning namespace {})", namespacePath.getName());
+    return Arrays.stream(fs.listStatus(namespacePath, this::isValidTableDirectory))
+      .parallel()
+      .filter(FileStatus::isDirectory)
+      .flatMap(tableDir -> {
+        try {
+          List<Path> regionDirs = FSUtils.getRegionDirs(fs, tableDir.getPath());
+          List<RegionInfo> tableRegions = scanRegionsInTable(fs, regionDirs);
+          LOG.debug("Found {} regions in table {} in namespace {}",
+            tableRegions.size(), tableDir.getPath().getName(), namespacePath.getName());
+          return tableRegions.stream();
+        } catch (IOException e) {
+          LOG.warn("Failed to scan table directory: {} for namespace {}",
+            tableDir.getPath(), namespacePath.getName(), e);
+          return Stream.empty();
+        }
+      }).toList();
+  }
+
+  private List<RegionInfo> scanRegionsInTable(FileSystem fs, List<Path> regionDirs) throws IOException {
+    List<RegionInfo> regions = new ArrayList<>();
+
+    for (Path regionDir : regionDirs) {
+      String encodedRegionName = regionDir.getName();
+      try {
+        Path regionInfoPath = new Path(regionDir, HRegionFileSystem.REGION_INFO_FILE);
+        if (fs.exists(regionInfoPath)) {
+          RegionInfo ri = readRegionInfo(fs, regionInfoPath);
+          if (ri != null && isValidRegionInfo(ri, encodedRegionName)) {
+            regions.add(ri);
+            LOG.debug("Found region: {} -> {}", encodedRegionName, ri.getRegionNameAsString());
+          } else {
+            LOG.warn("Invalid RegionInfo in file: {}", regionInfoPath);
+          }
+        } else {
+          LOG.debug("No .regioninfo file found in region directory: {}", regionDir);
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to read region info from directory: {}", encodedRegionName, e);
+      }
+    }
+    return regions;
+  }
+
+  private boolean isValidTableDirectory(Path path) {
+    return !(path.getName().matches("^[._-].*")) &&
+      !(path.getName().startsWith(TableName.META_TABLE_NAME.getNameAsString()));
+  }
+
+  private boolean isValidRegionInfo(RegionInfo regionInfo, String expectedEncodedName) {
```

Comment:
Same here. No HBase standard tool for that?

Reply:
This way of validation is consistent with other classes like HBaseFsck/ClientMetaTableAccessor/MobUtils etc. HBaseFsck is deprecated and in the rest, it was mostly in line validation. So I just kept this method as is.